src/android/Dgram.java

```diff
@@ -75,198 +48,191 @@ public void run() {
             DatagramPacket datagramPacket = new DatagramPacket(data, data.length);
             while (true) {
                 try {
-                    if (this.m_datagramSocket.isClosed()) {
+                    if (this.datagramSocket.isClosed()) {
                         Log.d(TAG, "Exiting message loop because socket is closed.");
                         return;
                     }
 
                     // Reset the length in case we receive an incomplete DatagramPacket
                     datagramPacket.setLength(data.length);
-                    this.m_datagramSocket.receive(datagramPacket);
+                    this.datagramSocket.receive(datagramPacket);
                     String message = new String(data, 0, datagramPacket.getLength(), "UTF-8");
                     String address = datagramPacket.getAddress().getHostAddress();
                     int port = datagramPacket.getPort();
                     Log.d(TAG, "Received message " + message + " from " + address);
-                    Dgram.this.m_javascriptQueue.put("_onMessage("
-                            + this.m_datagramSocketId + ","
-                            + "'" + message + "',"
-                            + "'" + address + "',"
-                            + port + ");");
-                } catch (InterruptedException e) {
-                    Log.d(TAG, e.toString());
-                } catch (Exception e) {
-                    Log.d(TAG, "Receive exception:" + e.toString());
-
-                    try {
-                        Dgram.this.m_javascriptQueue.put("_onError("
-                                + this.m_datagramSocketId
-                                + ",'receive','"
-                                + e.toString() + "');");
-                    } catch (Exception innerException) {
-                        Log.e(TAG, "Receive exception:" + innerException.toString());
-                    } finally {
-                        return;
+                    if (onMessageCallback != null) {
+                        sendMessageResult(message, address, port);
                     }
+                } catch (Exception e) {
+                    Log.d(TAG, "Received exception:" + e.toString());
+                    sendMessageErrorResult(e);
                 }
             }
         }
     }
 
+    private void sendMessageResult(String message, String address, int port) {
+        JSONObject payload = new JSONObject();
+
+        try {
+            payload.put("message", message);
+            payload.put("address", address);
+            payload.put("port", port);
+            CallbackUtil.sendPluginResult(onMessageCallback, new PluginResult(PluginResult.Status.OK, payload));
+        } catch (Exception e) {
+            PluginResult result = new PluginResult(
+                    PluginResult.Status.ERROR,
+                    "Error occurred while creating onMessage payload: " + e.toString()
```

Comment:
i am trying to understand the error case here.  the only thing that could fail is the `CallbackUtil.sendPluginResult` and we then immediately try to use it again.